File: IAIso-v5.0/core/iaiso-php/src/Core/PressureEngine.php
<?php

declare(strict_types=1);

namespace IAIso\Core;

use IAIso\Audit\Event;
use IAIso\Audit\NullSink;
use IAIso\Audit\Sink;

/**
 * The IAIso pressure engine. Tracks accumulated load, decays over time,
 * and emits lifecycle events on threshold crossings.
 *
 * <p>See {@code spec/pressure/README.md} for normative semantics.
 *
 * <p>Wire-format events emitted: {@code engine.init}, {@code engine.step},
 * {@code engine.step.rejected}, {@code engine.escalation},
 * {@code engine.release}, {@code engine.locked}, {@code engine.reset}.
 */
final class PressureEngine
{
    private readonly PressureConfig $cfg;
    private readonly string $executionId;
    private readonly Sink $audit;
    private readonly Clock $clock;
    private readonly Clock $timestampClock;

    // mutable state
    private float $pressure = 0.0;
    private int $stepCount = 0;
    private Lifecycle $lifecycle = Lifecycle::Init;
    private float $lastStepAt = 0.0;

    public function __construct(PressureConfig $cfg, EngineOptions $opts)
    {
        $cfg->validate();
        $this->cfg = $cfg;
        $this->executionId = $opts->executionId !== ''
            ? $opts->executionId
            : 'exec-' . bin2hex(random_bytes(6));
        $this->audit = $opts->auditSink ?? NullSink::instance();
        $this->clock = $opts->clock ?? WallClock::instance();
        $this->timestampClock = $opts->timestampClock ?? $this->clock;
        $this->lastStepAt = $this->clock->now();
        $this->emit('engine.init', ['pressure' => 0.0]);
    }

    public function getConfig(): PressureConfig { return $this->cfg; }
    public function getExecutionId(): string    { return $this->executionId; }
    public function getPressure(): float        { return $this->pressure; }
    public function getLifecycle(): Lifecycle   { return $this->lifecycle; }

    public function snapshot(): PressureSnapshot
    {
        return new PressureSnapshot($this->pressure, $this->stepCount, $this->lifecycle, $this->lastStepAt);
    }

    /** Account for a unit of work; advance the engine. */
    public function step(StepInput $work): StepOutcome
    {
        if ($this->lifecycle === Lifecycle::Locked) {
            $this->emit('engine.step.rejected', [
                'reason'           => 'locked',
                'requested_tokens' => $work->tokens,
                'requested_tools'  => $work->toolCalls,
            ]);
            return StepOutcome::Locked;
        }

        $now = $this->clock->now();
        $elapsed = max(0.0, $now - $this->lastStepAt);

        $delta = ($work->tokens / 1000.0) * $this->cfg->tokenCoefficient
            + $work->toolCalls * $this->cfg->toolCoefficient
            + $work->depth * $this->cfg->depthCoefficient;
        $decay = $this->cfg->dissipationPerStep + $elapsed * $this->cfg->dissipationPerSecond;

        $this->pressure = self::clamp01($this->pressure + $delta - $decay);
        $this->stepCount += 1;
        $this->lastStepAt = $now;
        $this->lifecycle = Lifecycle::Running;

        $this->emit('engine.step', [
            'step'       => $this->stepCount,
            'pressure'   => $this->pressure,
            'delta'      => $delta,
            'decay'      => $decay,
            'tokens'     => $work->tokens,
            'tool_calls' => $work->toolCalls,
            'depth'      => $work->depth,
            'tag'        => $work->tag,
        ]);

        $pressure = $this->pressure;
        if ($pressure >= $this->cfg->releaseThreshold) {
            $this->emit('engine.release', [
                'pressure'  => $pressure,
                'threshold' => $this->cfg->releaseThreshold,
            ]);
            $this->pressure = 0.0;
            if ($this->cfg->postReleaseLock) {
                $this->lifecycle = Lifecycle::Locked;
                $this->emit('engine.locked', ['reason' => 'post_release_lock']);
            } else {
                $this->lifecycle = Lifecycle::Running;
            }
            return StepOutcome::Released;
        }
        if ($pressure >= $this->cfg->escalationThreshold) {
            $this->lifecycle = Lifecycle::Escalated;
            $this->emit('engine.escalation', [
                'pressure'  => $pressure,
                'threshold' => $this->cfg->escalationThreshold,
            ]);
            return StepOutcome::Escalated;
        }
        return StepOutcome::Ok;
    }

    /** Reset the engine. Emits {@code engine.reset}. */
    public function reset(): PressureSnapshot
    {
        $this->pressure = 0.0;
        $this->stepCount = 0;
        $this->lastStepAt = $this->clock->now();
        $this->lifecycle = Lifecycle::Init;
        $this->emit('engine.reset', ['pressure' => 0.0]);
        return $this->snapshot();
    }

    private function emit(string $kind, array $data): void
    {
        $this->audit->emit(new Event(
            $this->executionId,
            $kind,
            $this->timestampClock->now(),
            $data,
        ));
    }

    private static function clamp01(float $v): float
    {
        if ($v < 0.0) return 0.0;
        if ($v > 1.0) return 1.0;
        return $v;
    }
}

File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoCore/PressureEngine.swift
import Foundation
import IAIsoAudit

/// The IAIso pressure engine. Tracks accumulated load, decays over time,
/// and emits lifecycle events on threshold crossings.
///
/// See `spec/pressure/README.md` for normative semantics.
///
/// Wire-format events emitted: `engine.init`, `engine.step`,
/// `engine.step.rejected`, `engine.escalation`, `engine.release`,
/// `engine.locked`, `engine.reset`.
public final class PressureEngine: @unchecked Sendable {

    public let config: PressureConfig
    public let executionId: String

    private let audit: Sink
    private let clock: Clock
    private let timestampClock: Clock
    private let lock = NSLock()

    // Mutable state — guarded by `lock`.
    private var _pressure: Double = 0
    private var _stepCount: Int64 = 0
    private var _lifecycle: Lifecycle = .init
    private var _lastStepAt: Double

    public init(config: PressureConfig, options: EngineOptions) throws {
        try config.validate()
        self.config = config
        self.executionId = options.executionId
        self.audit = options.auditSink
        self.clock = options.clock
        self.timestampClock = options.timestampClock
        self._lastStepAt = options.clock.now()
        self.emit(kind: "engine.init", data: ["pressure": .double(0)])
    }

    public var pressure: Double {
        lock.lock(); defer { lock.unlock() }
        return _pressure
    }

    public var lifecycle: Lifecycle {
        lock.lock(); defer { lock.unlock() }
        return _lifecycle
    }

    public func snapshot() -> PressureSnapshot {
        lock.lock(); defer { lock.unlock() }
        return PressureSnapshot(
            pressure: _pressure, step: _stepCount,
            lifecycle: _lifecycle, lastStepAt: _lastStepAt)
    }

    /// Account for a unit of work; advance the engine.
    @discardableResult
    public func step(_ work: StepInput = .empty) -> StepOutcome {
        lock.lock()

        // Locked state: reject without advancing.
        if _lifecycle == .locked {
            lock.unlock()
            emit(kind: "engine.step.rejected", data: [
                "reason": .string("locked"),
                "requested_tokens": .int(work.tokens),
                "requested_tools": .int(work.toolCalls),
            ])
            return .locked
        }

        let now = clock.now()
        let elapsed = max(0, now - _lastStepAt)
        let delta = (Double(work.tokens) / 1000.0) * config.tokenCoefficient
            + Double(work.toolCalls) * config.toolCoefficient
            + Double(work.depth) * config.depthCoefficient
        let decay = config.dissipationPerStep + elapsed * config.dissipationPerSecond

        _pressure = clamp01(_pressure + delta - decay)
        _stepCount += 1
        _lastStepAt = now
        _lifecycle = .running

        let stepData: [String: AnyJSON] = [
            "step": .int(_stepCount),
            "pressure": .double(_pressure),
            "delta": .double(delta),
            "decay": .double(decay),
            "tokens": .int(work.tokens),
            "tool_calls": .int(work.toolCalls),
            "depth": .int(work.depth),
            "tag": work.tag.map { .string($0) } ?? .null,
        ]
        let pressureNow = _pressure
        let releaseThr = config.releaseThreshold
        let escThr = config.escalationThreshold
        let postReleaseLock = config.postReleaseLock

        lock.unlock()
        emit(kind: "engine.step", data: stepData)

        if pressureNow >= releaseThr {
            emit(kind: "engine.release", data: [
                "pressure": .double(pressureNow),
                "threshold": .double(releaseThr),
            ])
            lock.lock()
            _pressure = 0
            if postReleaseLock {
                _lifecycle = .locked
                lock.unlock()
                emit(kind: "engine.locked", data: ["reason": .string("post_release_lock")])
            } else {
                _lifecycle = .running
                lock.unlock()
            }
            return .released
        }
        if pressureNow >= escThr {
            lock.lock()
            _lifecycle = .escalated
            lock.unlock()
            emit(kind: "engine.escalation", data: [
                "pressure": .double(pressureNow),
                "threshold": .double(escThr),
            ])
            return .escalated
        }
        return .ok
    }

    /// Reset the engine. Emits `engine.reset`.
    @discardableResult
    public func reset() -> PressureSnapshot {
        lock.lock()
        _pressure = 0
        _stepCount = 0
        _lastStepAt = clock.now()
        _lifecycle = .init
        let snap = PressureSnapshot(
            pressure: _pressure, step: _stepCount,
            lifecycle: _lifecycle, lastStepAt: _lastStepAt)
        lock.unlock()
        emit(kind: "engine.reset", data: ["pressure": .double(0)])
        return snap
    }

    private func emit(kind: String, data: [String: AnyJSON]) {
        audit.emit(Event(
            executionId: executionId,
            kind: kind,
            timestamp: timestampClock.now(),
            data: data))
    }

    private func clamp01(_ v: Double) -> Double {
        if v < 0 { return 0 }
        if v > 1 { return 1 }
        return v
    }
}
